pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -134,22 +134,6 @@ func validABPSessionKey(key *ttnpb.KeyEnvelope) bool {
 
 // Set implements NsEndDeviceRegistryServer.
 func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest) (*ttnpb.EndDevice, error) {
-	if ttnpb.HasAnyField(req.FieldMask.Paths, "session.dev_addr") && (req.EndDevice.Session == nil || req.EndDevice.Session.DevAddr.IsZero()) {
```

Comment:
The only reason why someone would want to set the `session.dev_addr` to zero value is to make the device model corrupted. I don't think we should allow this.
The set transaction should fail anyway due to invalid `dev_addr`, but we can(and should) just directly send a clear error to the user without even hitting the stores or doing anything else

Reply:
The field mask indicate what the caller knows about the entity being passed.

If `session == nil`, a field mask containing `session.dev_addr` is not wrong. It should be ineffective.

Sure, NS should avoid invalid state. And a `session.dev_addr` that is zero is invalid state, so that should be checked, but there's no need to return an error if `session == nil`.

Reply:
I assume you mean `session == nil` on the stored device. The solution is simple though - we should get the `session` if `session.dev_addr` is requested and do different checks in update and create functionality

Reply:
> If `session == nil`, a field mask containing `session.dev_addr` is not wrong. It should be ineffective.

That is not true - if `session == nil`, but `session.dev_addr` is specified - the `session.dev_addr` on the stored device is deleted(assuming `session.dev_addr` is stored)

Reply:
> That is not true - if `session == nil`, but `session.dev_addr` is specified - the `session.dev_addr` on the stored device is deleted(assuming `session.dev_addr` is stored)

Ineffective was specifically for create. For update, again, the field mask specifies what the caller knows. I.e. the caller knows `session.dev_addr` and `session == nil`, so the entire `session` should be reset.

Reply:
That's not how fieldmasks work, right?
If you specify `a.b`, `a.c` must not change.
From https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/FieldMask:
> A field mask in update operations specifies which fields of the
 targeted resource are going to be updated. The API is required
 to only change the values of the fields as specified in the mask
 and leave the others untouched. If a resource is passed in to
 describe the updated values, the API ignores the values of all
 fields not covered by the mask.

Also, please note that the registry does not validate `session.dev_addr` set by the user (at least now) - that happens in this RPC.

Reply:
Right, then make it ineffective. `session` will also be part of the field mask anyway.